swagger.yaml

```diff
@@ -5317,6 +5317,21 @@ definitions:
     properties:
       type:
         $ref: '#/definitions/platform_type'
+      external:
+        $ref: '#/definitions/platform_external'
+
+  platform_external:
+    type: object
+    x-go-custom-tag: gorm:"embedded;embeddedPrefix:platform_external_"
+    description: The configuration for the external platform type.
+    properties:
+      platform_name:
+        type: string
+        x-nullable: true
+        enum: ['Unknown', 'oci']
+        default: 'Unknown'
+        description: |-
+          Holds the infrastructure provider name.
```

Comment:
@gamli75 @eliorerz

Don't review the rest for the PR as it's WIP, I want your opinion on this API change for [platform external](https://github.com/openshift/enhancements/blob/master/enhancements/cloud-integration/infrastructure-external-platform-type.md).

This would be the API change we need to let the user customize the platform name in the installer config when choosing external platform type, this value would also be used to customize the assisted service behavior (if `platform_name` is `oci` then allow minimal ISO only).

I set the default to `Unknown` because it's the default in the [api](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L237).

On create and update requests, we would check that `platform.external` field is set only when `platform.type` is equal to `external`.
On update request, if `platform.type` is different from `external` we will reset `platform.external` to `nil`.

About `platform_name` field, I wonder if shouldn't be better to call this field `profile` as this value conditions `platform_name` value in installer config and changes the behavior of the assisted service? (in that case we would have a `generic` and an `oci` profile)

@avishayt do you have an opinion on this change? Don't hesitate to contact me as I guess you don't have all the context.

Reply:
I need to take a deeper look in it, but from a first glace, that structure of the API was changed after a lot of discussions. We used to have `vsphere` and `ovirt` under platform definition (the same as you are adding the external one) and we removed it.
Not sure what is the best approach here because (to my understanding) the `exteranl` platform type will not be represented in the `install-config.yaml`.
@avishayt @eranco74 

e.g. 
https://github.com/openshift/assisted-service/blob/cb48591dee6f39382362922d6354743574b7454b/swagger.yaml#L4666-L4678

Reply:
The question is if we want to consider "external" to be a platform and expose that (as in this PR), or expose "oci" as the platform (i.e., just adding "oci" to the enum of supported platforms).
The benefit of the API as in this PR is that it aligns more with openshift installer.  The downside is that those platforms aren't first-class citizens in AI.  For example, say one day we have 2 external platforms ext1 and ext2 (no relation to the file system).  ext1 has properties username and password, and ext2 has the property api_key.  How do we present that to the user in a way that's not confusing?  If they are first-class citizens, we could have each one's properties represented as in the swagger that @eliorerz pasted above for how we had vsphere and ovirt.

Reply:
>  Not sure what is the best approach here because (to my understanding) the exteranl platform type will not be represented in the install-config.yaml.

FYI, the [implementation is on-going](https://issues.redhat.com/browse/OCPCLOUD-2036)

>  ext1 has properties username and password, and ext2 has the property api_key

in the install config, the external platform takes only 2 parameters:
- the platform name, for telemetry purposes
- whether we want to enable the external platform or not (I think that by default, we want to enable it, otherwise it behaves like none platform)

Then, we expect the user to upload the manifests that will configure the cloud integration (the CCM) at installation time. I think most of the details for the platform will be specified in this step.

I see your point at making the platform first-class citizen, I'll explore it and see with the UI team what it means for them as I think they will have to know if a platform is external or not in order to tune the journey for the user (e.g.: guide them to the upload manifest screen).

Otherwise, the idea I had was to define a profile for the external platform, and use one more indirection to pass additional options, which  I agree, makes things a bit more complex:
```
platform_type: external
external:
  profile: ext1
  ext1:
    username: ...
    password: ...
```

Reply:
@jkilzi what would it mean on UI side if we define `oci` as a new platform type (under the hood the BE will use the external platform) ?

e.g.:
```
platform:
  type: oci
```

Reply:
That could work well, could save us passing also `platform.name: 'oci'`. So IIUC `oci` will be an alias for `external`?